### DISCUSSION TITLE
Call is not executed

### DISCUSSION THREAD

**Discussion by JacopoNeroni** [wevm/wagmi#2938] [Q&A]
In a react project where I am using wagmi, I call this function from a button:

```
const activeNos = () => {
        if(nosApproved === true){
            writeActiveNos?.();
        } else {
            writeApproveNosStacking?.();
        }
    }
```

If the approve has not been executed, it calls the approve function (else) and executes it without any problems. In the success of the approve function I modify a state 'nosApproved' and set it to true.
The problem occurs later, it correctly enters the loop if 'if(nosApproved === true)', but does not execute the function 'writeActiveNos?.();'. If I refresh the page then it works correctly. I have checked that the arguments I pass to the writeActiveNos function are correctly valued, I don't understand why it doesn't run. This is the code for the writeActiveNos function.

```
    const {config: configActiveNos} = usePrepareContractWrite({
        mode: 'recklesslyUnprepared',
        address: stackingContractAddress,
        abi: stackingABI,
        chainId: chainId,
        functionName: 'activeBoost',
        args: [tokensOwnedByNosData],
    });
    const { write: writeActiveNos } = useContractWrite({
        ...configActiveNos,
        onSuccess(dataActiveNos) {
            setNosStatus("Activated");
            setStakeActive(true);
        },
        onError(error) {
            console.log('Error', error)
        },
    });
```

**Comment by SaadNiaziDev** [comment]
When calling approve functions, they trigger onSuccess as soon as you accept the transaction. But in background, it takes 3-6s to complete the approve call,mine transaction and create allowance. The reason in your case could be similar.As soon as you call approve function and accept it, it will trigger onSuccess and your next function would be call but it will fail as allowance is not allocated yet.

Can you try a simple hack( just to see that what I am saying might be the problem) .
Do this in your approve function call.

`onSuccess(data) {
setTimeout(()=>{
            setNosApproved (true)}
,3000)
}`

**Reply by JacopoNeroni** [reply]
It still doesn't work, I don't think that's the problem.

**Reply by JacopoNeroni** [reply]
```
    //approve nos stacking
    const {config: configApproveNosStacking} = usePrepareContractWrite({
        mode: 'recklesslyUnprepared',
        address: nosContractAddress,
        abi: nosABI,
        chainId: chainId,
        functionName: 'approve',
        args: [stackingContractAddress, tokensOwnedByNosData],
    });
    const { write: writeApproveNosStacking } = useContractWrite({
        ...configApproveNosStacking,
        onSuccess(dataApproveNosStacking) {
            console.log('Success Approve Stacking', dataApproveNosStacking);
            setTimeout(()=>{ setNosApproved (true)} ,3000)
        },
    });
```
